openedx-unsupported/configuration
# Allows jenkins analytics to run analytics tasks on AWS resources

Allows jenkins analytics jobs to ssh to AWS resources and, using ssh key forwarding, clone secure repos.

Also fixes issue with re-building the edx-analytics-pipeline dependencies (detailed in (closed) [edx-analytics-pipeline PR #228](https://github.com/edx/edx-analytics-pipeline/pull/228)).
- Sets JENKINS_HOME to /edx/var/jenkins in `/etc/defaults/jenkins` instead of using the default `/var/lib/jenkins` symlink.  Accessing JENKINS_HOME via the symlink was causing the pyinstrument pip dependency to fail on re-install.  I've left the symlink in place though, because travis-ci failed without it.
- Creates a second credential (in addition to the git credential) for shelling into the AWS service.  This credential is pulled from a file in the configured secure-repo, which is cloned when the seed task is run.
- Adds missing package dependencies required to build the analytics-pipeline on the jenkins server.

**JIRA tickets**: [OLIVE-32](https://openedx.atlassian.net/browse/OLIVE-32), [OC-1512](https://tasks.opencraft.com/browse/OC-1512)

**Discussions**: See [discussion on PR 15](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/15#issuecomment-204265069).

**Dependencies**:
- [edx-analytics-configuration PR #30](https://github.com/edx/edx-analytics-configuration/pull/30) - fixes ansible 1.4.4 incompatibility
- [edx-jenkins-job-dsl PR #19](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/19) - wraps the shell commands with both sshAgent credentials: github, AWS
- [edx-analytics-pipeline PR #232](https://github.com/edx/edx-analytics-pipeline/pull/232) - fixes bug with initial clone of the secure repo
- [edx-jenkins-job-dsl PR #23](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/23) - fixes bug detected during this PR review, and adds unit tests for AnalyticsTasks.parseYaml

**Sandbox URL**: 

http://52.20.136.133:8080 - ping @pomegranited for login credentials.

**Testing instructions**:
1. Install edx/configuration requirements in a virtualenv:
   
   ```
   virtualenv venv
   . venv/bin/activate
   git clone https://git@github.com/edx/configuration.git
   cd configuration
   make requirements
   ```
2. Verify that you have access to https://github.com/open-craft/edx-jenkins-job-dsl/
3. Create a YAML file containing the jenkins analytics configuration.  See [jenkins_analytics README](https://github.com/edx/configuration/tree/master/playbooks/roles/jenkins_analytics) and/or [example ansible-jenkins-credentials.yml.zip](https://github.com/edx/edx-analytics-configuration/files/205786/ansible-jenkins-credentials.yml.zip) for details.
4. Build the vagrant instance, using the YAML file created above:
   
   ```
   cd vagrant/base/analytics_jenkins
   export VAGRANT_JENKINS_LOCAL_VARS_FILE=/path/to/your.yml
   vagrant up
   ```
   
   Expected result is that you will get a working jenkins instance that you'll be able to log in to 192.168.33.15:8080 or localhost:8080, using the `jenkins` user, and the password in `JENKINS_ANALYTICS_USER_PASSWORD_PLAIN`.  The AnalyticsSeedJob will have run successfully, and created an AnswerDistributionWorkflow job.
5. Via your AWS GUI, generate an Access Key for use by the vagrant VM.  URL looks something like this: https://console.aws.amazon.com/iam/home?#users/<your username>
6. Shell into your vagrant instance, and grant the jenkins user access to your AWS instance.
   
   ```
   vagrant ssh
   sudo su jenkins
   cd
   virtualenv aws-venv
   . aws-venv/bin/activate
   pip install awscli
   aws configure
   # AWS Access Key ID [None]: <your key>
   # AWS Secret Access Key [None]: <your code>
   # Default region name [None]: <your region>
   # Default output format [None]:
   ```
7. Via the Jenkins GUI, Build the AnswerDistributionWorkflow job with the configured parameters.

Expected result is the task will run successfully, and terminate the EMR at the end.  Previously, the analytics task would fail after EMR provisioning, because it was unable to clone the secure repo on the EMR instance.

Re-running the AnswerDistributionWorkflow will produce the same result.  Previously, every other analytics task build failed due to pyinstrument re-install errors.

**Reviewers**
- [ ] @jibsheet 

**Settings**
- See [jenkins_analytics README](https://github.com/open-craft/configuration/blob/jill/olive-22-jenkins-sshagent/playbooks/roles/jenkins_analytics/README.md) for details on the ansible variables.
- See [ansible-jenkins-credentials.yml.zip](https://github.com/edx/configuration/files/209720/ansible-jenkins-credentials.yml.zip) for example.
